src/mongo/db/storage/csv_file.cpp

```diff
@@ -0,0 +1,338 @@
+                                     const std::string& data) {
+    builder.append(fieldName, data);
+}
+
+template <>
+void appendTo<CsvFieldType::kDate>(BSONObjBuilder& builder,
+                                   const std::string& fieldName,
+                                   const std::string& data) {
+    auto date = dateFromISOString(data);
+    // return badValue error codes
+    uassert(date.getStatus().code(), date.getStatus().toString(), date.isOK());
+    builder.appendDate(fieldName, date.getValue());
+}
+
+template <>
+void appendTo<CsvFieldType::kOid>(BSONObjBuilder& builder,
+                                  const std::string& fieldName,
+                                  const std::string& data) {
+    constexpr size_t kLengthOidValue = 24;
+    constexpr size_t kOidTypeStr = 8;                      // objectid
+    constexpr size_t kOidTypeStrPrefix = kOidTypeStr + 2;  // objectid("
+
+    std::string mutableData = data;
+    if (data[0] == 'O' || data[0] == 'o') {
+        std::transform(
+            mutableData.begin(), mutableData.begin() + kOidTypeStr, mutableData.begin(), ::tolower);
+        uassert(ErrorCodes::BadValue,
+                "Invalid Object Id format",
+                mutableData.substr(0, kOidTypeStrPrefix) == "objectid(\"");
+        mutableData = mutableData.substr(kOidTypeStrPrefix, kLengthOidValue);
+    } else if (data[0] == '\"') {
+        mutableData = mutableData.substr(1, kLengthOidValue);
+    }
+
+    uassert(
+        ErrorCodes::BadValue, "Invalid Object Id Format", mutableData.length() == kLengthOidValue);
+    mongo::OID id = mongo::OID(mutableData);  // throws exception if bad
+    builder.append(fieldName, id);
+}
+
+template <>
+void appendTo<CsvFieldType::kBool>(BSONObjBuilder& builder,
+                                   const std::string& fieldName,
+                                   const std::string& data) {
+    std::string mutableData = data;
+    bool val;
+    std::transform(data.begin(), data.end(), mutableData.begin(), ::tolower);
+    if (mutableData == "true" || mutableData == "t" || mutableData == "yes" || mutableData == "y" ||
+        mutableData == "1")
+        val = true;
+    else if (mutableData == "false" || mutableData == "f" || mutableData == "no" ||
+             mutableData == "n" || mutableData == "0")
+        val = false;
+    else
+        uasserted(ErrorCodes::BadValue, "Invalid Boolean representation");
+    builder.append(fieldName, val);
+}
+
+
+boost::optional<BSONObj> CsvFileInput::readBsonObj() {
+    std::string record;
+    std::getline(_ifs, record);
+    // if eof is reache, _ifs return false
+    if (_ifs.eof() || _ifs.fail()) {
+        return boost::none;
+    }
+
+    auto data = parseLine(record);
+    uassert(ErrorCodes::TypeMismatch,
+            "Metadata and current record mismatch",
+            data.size() == _md.size());
+
+    BSONObjBuilder builder;
+
+    for (size_t i = 0; i < _md.size(); ++i) {
+        if (data[i] == "") {
```

Comment:
If the type is string, should we return null for an empty string or an empty string?

Reply:
Assume that the second field is a string type

1. A,,C
2. "A","","C"

Assume that the first field is a int32 type

1. ,B,C

We agreed that an empty value for a string type is interpreted as null.

{a: 1, b: null}
== {a: 1} ?